hw_classes/myfasta.py

```diff
@@ -0,0 +1,136 @@
+import seaborn as sns
+import matplotlib.pyplot as plt
+
+
+class Fasta():
+
+    """A class that that collects some statistics from fasta file"""
+
+    def __init__(self, path):
+        self.path = path
+        self.seq_dict = {}
+        counter = 0
+        with open(path) as file:
+            name = ''
+            seq = ''
+            for line in file:
+                counter += 1
+                line = line.strip()
+                if line == '':
+                    pass
+                if line.startswith('>'):
+                    if counter > 1:
+                        self.seq_dict[name] = seq
+                        seq = ''
+                        name = line[1:]
+                    else:
+                        seq = ''
+                        name = line[1:]
+                else:
+                    seq = seq + line
+        self.seq_dict[name] = seq
+
+    def count_seqs(self):
+        return len(self.seq_dict)
+
+    def plot_len(self, save=False, name='./length.png'):
+        lengths = [len(value) for value in self.seq_dict.values()]
+        ax = sns.histplot(x=lengths, kde=True)
+        ax.set(xlabel='length', title='Length histogram')
+        if save:
+            plt.savefig(fname=name, bbox_inches='tight')
+        plt.show()
+        plt.close()
+
+    def gc(self):
+        try:
+            self.gc_dict
+            return (self.gc_dict)
+        except AttributeError:
+            self.gc_dict = {}
+            for key, value in self.seq_dict.items():
+                seq = value.casefold()
+                gc_all = seq.count('g')+seq.count('c')
+                atgc_all = seq.count('g')+seq.count('c')+seq.count('a')+seq.count('t')
+                gc_content = gc_all/atgc_all
+                self.gc_dict[key] = gc_content
+            return self.gc_dict

```

Comment:
Мы не рассказывали, но для таких случаев в питоне есть встроенная функция `getattr`, которая позволяет получить значение атрибута у какого-нибудь объекта или дефолтное, если его нет.
```suggestion
    def gc(self):
        gc_dict = getattr(self, "gc_dict", None)
        if gc_dict:
            return gc_dict
        else:
            self.gc_dict = {}
            for key, value in self.seq_dict.items():
                seq = value.casefold()
                gc_all = seq.count('g')+seq.count('c')
                atgc_all = seq.count('g')+seq.count('c')+seq.count('a')+seq.count('t')
                gc_content = gc_all/atgc_all
                self.gc_dict[key] = gc_content
            return self.gc_dict

```